src/persist-client/src/impl/machine.rs

```diff
@@ -263,11 +273,16 @@ where
     pub async fn downgrade_since(
         &mut self,
         reader_id: &ReaderId,
+        outstanding_seqno: Option<SeqNo>,
         new_since: &Antichain<T>,
         heartbeat_timestamp_ms: u64,
     ) -> (SeqNo, Since<T>, RoutineMaintenance) {
         let metrics = Arc::clone(&self.metrics);
         self.apply_unbatched_idempotent_cmd(&metrics.cmds.downgrade_since, |seqno, state| {
+            let seqno = match outstanding_seqno {
+                Some(outstanding_seqno) => std::cmp::min(outstanding_seqno, seqno),
```

Comment:
the blind std::cmp::min makes me nervous because calling it incorrectly could break internal persist invariants about seqno_since not going backward. I think this wants to assert that outstanding_seqno, if some, is >= the readers current seqno capability in state to prevent misuse

also, push this logic into the state.downgrade-since call, please